crates/core/src/constraint_system/prove.rs

```diff
@@ -479,86 +479,148 @@ fn make_masked_flush_witnesses<'a, U, Tower>(
+						let nonzero_scalars_prefixes = (1 << oracles.n_vars(*id)) - zero_suffix;
 
-				let polys = inner_polys
-					.iter()
-					.map(|id| witness.get_multilin_poly(*id))
-					.collect::<Result<Vec<_>, _>>()?;
+						witness.update_multilin_poly_with_nonzero_scalars_prefixes([(
+							*id,
+							poly,
+							nonzero_scalars_prefixes,
+						)])?;
+					}
 
-				let n_vars = composite.n_vars();
-				let log_width = <PackedType<U, FExt<Tower>>>::LOG_WIDTH;
+					let polys = inner_polys
+						.iter()
+						.map(|id| witness.get_multilin_poly(*id))
+						.collect::<Result<Vec<_>, _>>()?;
 
-				let packed_len = 1 << n_vars.saturating_sub(log_width);
+					let max_packed_zero_suffix =
+						zero_suffixes.into_iter().max().unwrap_or(0) >> log_width;
 
-				let inner_c = composite.c();
+					let max_packed_zero_suffix = max_packed_zero_suffix.min(packed_len);
 
-				let composite_data = (0..packed_len)
-					.into_par_iter()
-					.map(|i| {
-						<PackedType<U, FExt<Tower>>>::from_fn(|j| {
-							let index = i << <PackedType<U, FExt<Tower>>>::LOG_WIDTH | j;
+					let mut composite_data = (0..packed_len - max_packed_zero_suffix)
+						.into_par_iter()
+						.map(|i| {
 							let evals = polys
 								.iter()
-								.map(|poly| poly.evaluate_on_hypercube(index).unwrap_or_default())
+								.map(|poly| {
+									<PackedType<U, FExt<Tower>>>::from_fn(|j| {
+										let index =
+											i << <PackedType<U, FExt<Tower>>>::LOG_WIDTH | j;
+										poly.evaluate_on_hypercube(index).unwrap_or_default()
+									})
+								})
 								.collect::<Vec<_>>();
 
 							inner_c
 								.evaluate(&evals)
 								.expect("query length is the same as poly length")
 						})
-					})
-					.collect::<Vec<_>>();
-
-				let composite_poly = MultilinearExtension::new(n_vars, composite_data)
-					.expect("data is constructed with the correct length with respect to n_vars");
-
-				Ok((flush_oracle, MLEDirectAdapter::from(composite_poly).upcast_arc_dyn()))
-			}
-			MultilinearPolyVariant::LinearCombination(lincom) => {
-				let polys = lincom
-					.polys()
-					.map(|id| witness.get_multilin_poly(id))
-					.collect::<Result<Vec<_>, _>>()?;
-
-				let packed_len = 1
-					<< lincom
-						.n_vars()
-						.saturating_sub(<PackedType<U, FExt<Tower>>>::LOG_WIDTH);
-				let lin_comb_data = (0..packed_len)
-					.into_par_iter()
-					.map(|i| {
-						<PackedType<U, FExt<Tower>>>::from_fn(|j| {
-							let index = i << <PackedType<U, FExt<Tower>>>::LOG_WIDTH | j;
-							polys.iter().zip(lincom.coefficients()).fold(
-								lincom.offset(),
-								|sum, (poly, coeff)| {
-									sum + poly
-										.evaluate_on_hypercube_and_scale(index, coeff)
-										.unwrap_or(<FExt<Tower>>::ZERO)
-								},
-							)
+						.collect::<Vec<_>>();
+
+					// `ArithExpr::Const(F::ONE) + selector * arith_expr_linear` — so if selector is
+					// zero, we fill with ones.
+					if n_vars < PackedType::<U, FExt<Tower>>::LOG_WIDTH {
```

Comment:
What's the top branch for? It should be OK to make the multilinear extension with extra 1s at the end, even if `n_vars < LOG_WIDTH`.